### PR TITLE
Subrecord Types

### PR DESCRIPTION
**Pull request by zakird**
This adds types as subclasses.

So now you can do something like this:
```
SSH = SubRecordType({
   "banner":SubRecord({
       "comment":String(),
       "timestamp":DateTime()
       })
   })

Record({
    "ssh":SSH(doc="instance doc", required=True)
})
```

### CODE REVIEW

**Comment by zakird** [comment]
OK @justinbastress I think this is ready for review. I do agree that there's a bit higher learning curve here, but we don't break any schemas, so people can move forward as they see fit.

File: zschema/compounds.py [hunk 1]
```diff
@@ -207,6 +209,22 @@ def validate(self, name, value):
...
+        exclude=None,
+        category=None):
+    attrs = {
+        "definition":definition,
+        "required":required,
+        "allow_unknown":allow_unknown,
+        "_exclude":exclude if exclude else set([]),
+        "category":category
+    }
+    return type("SubRecord", (SubRecord,), attrs)
```
[the remark this answers is not in the document]
Comment (zakird) [reply zschema/compounds.py]: agreed. We can fix this is another PR though. It's not particularly related to this change.

**Comment by zakird** [comment]
I'll note, there was one test case that used `.new()`. It was trivially replaceable with the new functionality, so I did so.

File: zschema/tests.py [hunk 2]
```diff
@@ -569,6 +569,30 @@ def test_parses_ipv4_records(self):
...
+        ssh = SSH(doc="instance doc", required=True)
+        ipv4_host_ssh = Record({
+            Port(22):SubRecord({
+                "ssh":ssh
+            })
+        })
+        self.assertEqual(ssh.doc, "instance doc")
+        self.assertEqual(ssh.required, True)
+        ipv4_host_ssh.validate(json_fixture('ipv4-ssh-record'))
+        # class unchanged
```
Comment (justinbastress) [inline zschema/tests.py]: Note -- no test for inheriting the class attributes in the child

**Comment by cdzombak** [comment]
> I do agree that there's a bit higher learning curve here, but we don't break any schemas, so people can move forward as they see fit.

@zakird can you please clarify what you mean by "we don't break any schemas?" This is a backward-incompatible change if any zcshema users are using `.new()`, correct?

**Comment by zakird** [comment]
This is true and will require a major revision to be released. I think it's
unlikely to have been used any places. I looked through schemas I know
(e.g., ones in ZGRAB, ztag, darknet processing) and it doesn't look invoked
anywhere.

This probably isn't too surprising as it didn't have much use until this
week and that was never part of a release, so I don't think should be an
issue to remove from that standpoint.

If we want to be super careful, we could mark the old version deprecated
today and remove later.

**Comment by justinbastress** [comment]
Another item -- any idea how this could be naturally extended to `ListOf()`? A parallel implementation of `ListOfType()` would probably work, though the name isn't as clear as `SubRecordType`.

**Comment by zakird** [comment]
Offhand, I feel like that the construction of ListOf should feel the same, and just take in things like doc and required. I think it should have to take in a field not a type. 

Yes, I agree, types should be capital letters.

**Comment by zakird** [comment]
Also thinking more, absolutely agree with you @cdzombak on taking the more responsible approach in .new. I'll mark it as deprecated in this PR and then we can altogether remove in a future release.

**Comment by justinbastress** [comment]
> Offhand, I feel like that the construction of ListOf should feel the same, and just take in things like doc and required. I think it should have to take in a field not a type.

Yeah, `ListOf` shouldn't change and should always take in an instance. But, it would be nice if lists could be treated the same way as records, e.g.
```python
Certificate = SubRecordType({...})
CertificateChain = ListOfType(Certificate())
# ...
scan_result = SubRecord({
    "certificate": Certificate(doc="...", required=True),
    "intermediates": CertificateChain(doc="...", required=True),
})
```

rather than
```python
Certificate = SubRecordType({...})
certificate_chain = ListOf(Certificate())
# ...
scan_result = SubRecord({
    "certificate": Certificate(doc="...", required=True),
    "intermediates": certificate_chain, # can't set docs/required
})
```

As it stands, `ListOf(Certificates(), doc="...")` could be used inline each time rather than using the `certificate_chain` alias, but that means that you have to know the exact type of `certificate_chain` everywhere it's used, and if it changes, all instances need to be tracked down and updated.

File: zschema/leaves.py [hunk 3]
```diff
@@ -319,31 +319,20 @@ def _validate(self, name, value):
                     name, str(value), str(min_)))
 
 
-class Integer(Signed32BitInteger):
-
-    DEPRECATED = True
-
-
-class Signed8BitInteger(Integer):
+class Signed8BitInteger(Leaf):
```
[the remark this answers is not in the document]
Comment (zakird) [reply zschema/leaves.py]: Gah OK. I thought tests would have caught that. I'll fix and add a test.

File: zschema/tests.py [hunk 4]
```diff
@@ -615,10 +667,10 @@ def test_get_registered(self):
...
     def test_subrecord_child_types_can_override_parent_attributes(self):
-        Certificate = SubRecord({}, doc="A parsed certificate.")
+        Certificate = SubRecordType({}, doc="A parsed certificate.")
         OtherType = SubRecord({
-            "ca": Certificate.new(doc="The CA certificate."),
-            "host": Certificate.new(doc="The host certificate."),
+            "ca": Certificate(doc="The CA certificate."),
+            "host": Certificate(doc="The host certificate."),
         })
         self.assertEqual("A parsed certificate." , Certificate.doc)
```
Comment (justinbastress) [inline zschema/tests.py]: To be identical with the old test, it should actually be ` self.assertEqual("A parsed certificate." , Certificate().doc)`. The current behavior is the same check as done on line 619.

**Comment by justinbastress** [comment]
This now runs against my updated schemas.

**Comment by zakird** [comment]
@cdzombak you cool with merging?

**Comment by cdzombak** [comment]
> you cool with merging?

I guess so, though there seem to be a lot of changes here, which makes me nervous.

This merge will cause our schema (https://github.com/zmap/ztag/blob/master/ztag/schema.py) to start printing deprecation warnings, so someone will need to address that in a future ztag PR.

**Pull request merged**